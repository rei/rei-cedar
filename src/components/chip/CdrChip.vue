<script>
import { defineComponent, useCssModule } from 'vue';

export default defineComponent({
  name: 'CdrChip',
  setup() {
    return {
      style: useCssModule(),
      baseClass: 'cdr-chip',
    };
  },
});
</script>

<template>
  <button
    :class="style[baseClass]"
  >
    <span
      v-if="$slots['icon-left']"
      :class="style['cdr-chip__icon-left']"
    >
      <slot name="icon-left" />
    </span>
    <span
      v-if="$slots['icon-right']"
      :class="style['cdr-chip__icon-right']"
    >
      <slot name="icon-right" />
    </span>
    <span :class="style['cdr-chip__content']">
      <slot />
    </span>
  </button>
</template>

<style lang="scss" module src="./styles/CdrChip.module.scss">
</style>
