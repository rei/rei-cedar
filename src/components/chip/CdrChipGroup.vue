<script>
import {
  defineComponent, useCssModule, computed, ref, onMounted,
} from 'vue';

export default defineComponent({
  name: 'CdrChipGroup',
  props: {
    label: {
      type: String,
      required: true,
    },
    hideLabel: {
      type: Boolean,
      default: true,
    },
  },

  setup(props) {
    const baseClass = 'cdr-chip-group';
    const chipsEl = ref(null);

    const chips = ref([]);
    const currentIdx = ref(0);

    const nextIdx = computed(() => {
      const idx = currentIdx.value + 1;
      return idx >= chips.value.length ? 0 : idx;
    });
    const prevIdx = computed(() => {
      const idx = currentIdx.value - 1;
      return idx <= -1 ? (chips.value.length - 1) : idx;
    });
    const legendClass = computed(() => (props.hideLabel
      ? 'cdr-chip-group__legend--hidden'
      : 'cdr-chip-group__legend'));

    onMounted(() => {
      chips.value = Array.prototype.filter.call(
        chipsEl.value.children,
        (chip) => !(chip.getAttribute('disabled') === '' || chip.getAttribute('aria-disabled')),
      );
      currentIdx.value = Array.prototype.findIndex.call(
        chips,
        (chip) => chip.getAttribute('aria-checked') === 'true',
      );
    });

    const handleKeyDown = (e) => {
      // something besides the button is focused
      if (currentIdx.value === -1) return;

      const { key } = e;
      switch (key) {
        case 'Home':
          e.preventDefault();
          chips.value[0].focus();
          break;
        case 'End':
          e.preventDefault();
          chips.value[chips.value.length - 1].focus();
          break;
        case 'ArrowDown':
        case 'Down':
          e.preventDefault();
          chips.value[nextIdx.value].focus();
          break;
        case 'ArrowUp':
        case 'Up':
          e.preventDefault();
          chips.value[prevIdx.value].focus();
          break;
        default: break;
      }
    };
    const handleFocusIn = (e) => {
      // find out which, if any, button is focused
      currentIdx.value = Array.prototype.indexOf.call(chips.value, e.target);
    };
    return {
      style: useCssModule(),
      baseClass,
      chipsEl,
      legendClass,
      handleKeyDown,
      handleFocusIn,
    };
  },
});

</script>

<template>
  <fieldset
    :class="style[baseClass]"
    @focusin="handleFocusIn"
    @keydown="handleKeyDown"
  >
    <legend :class="style[legendClass]">
      <slot name="label">
        {{ label }}
      </slot>
    </legend>
    <div
      ref="chipsEl"
      :class="style['cdr-chip-group__content']"
    >
      <slot />
    </div>
  </fieldset>
</template>

<style lang="scss" module src="./styles/CdrChipGroup.module.scss">
</style>
