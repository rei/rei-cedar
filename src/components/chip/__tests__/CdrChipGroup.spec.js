import { h } from 'vue';
import { mount } from '../../../../test/vue-jest-style-workaround.js';

import CdrChipGroup from '../CdrChipGroup.vue';
import CdrChip from '../CdrChip.vue';

describe('CdrChipGroup', () => {
  let wrapper;
  let chips
  describe('default configuration', () => {
    beforeEach(() => {
      wrapper = mount(CdrChipGroup, {
        props: {
          label: 'test',
        },
        slots: {
          default: [
            h(CdrChip, { 'aria-checked': 'true', tabindex: 0, role: 'radio' }, { default: () => 'chip 1' }),
            h(CdrChip, { 'aria-checked': 'false', tabindex: -1, role: 'radio' }, { default: () => 'chip 2' }),
            h(CdrChip, { 'aria-checked': 'false', tabindex: -1, role: 'radio' }, { default: () => 'chip 3' }),
          ],
        },
        attachTo: document.body
      })
      chips = wrapper.findAll('.cdr-chip').map(domWrapper => domWrapper.element);
    })
    
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('keyboard interactions', () => {
      it('hitting ArrowUp focuses on the expected element', async () => {
        chips[0].focus();
        wrapper.trigger('keydown', { key: 'ArrowUp' });
        await wrapper.vm.$nextTick();
        expect(chips[2]).toBe(document.activeElement);
      });

      it('hitting ArrowDown focuses on the expected element', async () => {
        chips[2].focus();
        wrapper.trigger('keydown', { key: 'ArrowDown' });
        await wrapper.vm.$nextTick();
        expect(chips[0]).toBe(document.activeElement);
      });

      it('hitting End focuses on the expected element', async () => {
        chips[0].focus();
        wrapper.trigger('keydown', { key: 'End' });
        await wrapper.vm.$nextTick();
        expect(chips[2]).toBe(document.activeElement);
      });

      it('hitting Home focuses on the expected element', async () => {
        chips[2].focus();
        wrapper.trigger('keydown', { key: 'Home' });
        await wrapper.vm.$nextTick();
        expect(chips[0]).toBe(document.activeElement);
      });
    })
  });

  describe('when label is visible', () => {
    beforeEach(() => {
      wrapper = mount(h(CdrChipGroup, {
        label: 'test',
        hideLabel: false,
      },
        {
          default: () => [
            h(CdrChip, { 'aria-checked': true, tabindex: 0, role: 'radio' }, { default: () => 'chip 1' }),
            h(CdrChip, { 'aria-checked': false, tabindex: -1, role: 'radio' }, { default: () => 'chip 2' }),
          ]
        },
      ));
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('with label slot', () => {
    beforeEach(() => {
      wrapper = mount(h(CdrChipGroup, {
        label: 'test',
        hideLabel: false,
      },
        {
          default: () => [
            h(CdrChip, { 'aria-checked': true, tabindex: 0, role: 'radio' }, { default: () => 'chip 1' }),
            h(CdrChip, { 'aria-checked': false, tabindex: -1, role: 'radio' }, { default: () => 'chip 2' }),
          ],
          label: 'hey im overriding here!!!',
        },
      ));
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when current index is set', () => {
    beforeEach(() => {
      wrapper = mount(CdrChipGroup, {
        props: {
          label: 'test',
        },
        slots: {
          default: [
            h(CdrChip, { 'aria-checked': 'false', tabindex: -1, role: 'radio' }, { default: () => 'chip 1' }),
            h(CdrChip, { 'aria-checked': 'true', tabindex: 0, role: 'radio' }, { default: () => 'chip 2' }),
          ],
        },
      },
      );
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
