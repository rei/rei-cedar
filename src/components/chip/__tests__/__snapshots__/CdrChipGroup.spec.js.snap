// Vitest Snapshot v1

exports[`CdrChipGroup > default configuration > renders correctly 1`] = `
<fieldset
  class="cdr-chip-group"
>
  <legend
    class="cdr-chip-group__legend--hidden"
  >
    
    test
    
  </legend>
  <div
    class="cdr-chip-group__content"
  >
    
    <button
      aria-checked="true"
      class="cdr-chip"
      role="radio"
      tabindex="0"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 1
        
      </span>
    </button>
    <button
      aria-checked="false"
      class="cdr-chip"
      role="radio"
      tabindex="-1"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 2
        
      </span>
    </button>
    <button
      aria-checked="false"
      class="cdr-chip"
      role="radio"
      tabindex="-1"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 3
        
      </span>
    </button>
    
  </div>
</fieldset>
`;

exports[`CdrChipGroup > when current index is set > renders correctly 1`] = `
<fieldset
  class="cdr-chip-group"
>
  <legend
    class="cdr-chip-group__legend--hidden"
  >
    
    test
    
  </legend>
  <div
    class="cdr-chip-group__content"
  >
    
    <button
      aria-checked="false"
      class="cdr-chip"
      role="radio"
      tabindex="-1"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 1
        
      </span>
    </button>
    <button
      aria-checked="true"
      class="cdr-chip"
      role="radio"
      tabindex="0"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 2
        
      </span>
    </button>
    
  </div>
</fieldset>
`;

exports[`CdrChipGroup > when label is visible > renders correctly 1`] = `
<fieldset
  class="cdr-chip-group"
>
  <legend
    class="cdr-chip-group__legend"
  >
    
    test
    
  </legend>
  <div
    class="cdr-chip-group__content"
  >
    
    <button
      aria-checked="true"
      class="cdr-chip"
      role="radio"
      tabindex="0"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 1
        
      </span>
    </button>
    <button
      aria-checked="false"
      class="cdr-chip"
      role="radio"
      tabindex="-1"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 2
        
      </span>
    </button>
    
  </div>
</fieldset>
`;

exports[`CdrChipGroup > with label slot > renders correctly 1`] = `
<fieldset
  class="cdr-chip-group"
>
  <legend
    class="cdr-chip-group__legend"
  >
    
    hey im overriding here!!!
    
  </legend>
  <div
    class="cdr-chip-group__content"
  >
    
    <button
      aria-checked="true"
      class="cdr-chip"
      role="radio"
      tabindex="0"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 1
        
      </span>
    </button>
    <button
      aria-checked="false"
      class="cdr-chip"
      role="radio"
      tabindex="-1"
    >
      <!--v-if-->
      <!--v-if-->
      <span
        class="cdr-chip__content"
      >
        
        chip 2
        
      </span>
    </button>
    
  </div>
</fieldset>
`;
